File: backend/app/schemas/social.py
"""
Social Hub Pydantic Schemas
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum

from app.schemas.common import PaginatedResponse


class SocialPlatform(str, Enum):
    FACEBOOK = "facebook"
    INSTAGRAM = "instagram"
    LINKEDIN = "linkedin"
    TWITTER = "twitter"
    YOUTUBE = "youtube"
    TIKTOK = "tiktok"


class PostStatus(str, Enum):
    DRAFT = "draft"
    SCHEDULED = "scheduled"
    PUBLISHED = "published"
    FAILED = "failed"
    CANCELLED = "cancelled"


class PostType(str, Enum):
    TEXT = "text"
    IMAGE = "image"
    VIDEO = "video"
    CAROUSEL = "carousel"
    STORY = "story"
    REEL = "reel"


class SocialAccountResponse(BaseModel):
    """Social media account response"""
    id: str
    platform: SocialPlatform
    account_name: str
    account_id: str
    access_token: Optional[str] = None
    refresh_token: Optional[str] = None
    is_active: bool
    last_sync: Optional[datetime] = None
    follower_count: Optional[int] = None
    following_count: Optional[int] = None
    post_count: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class SocialPostResponse(BaseModel):
    """Social media post response"""
    id: str
    account_id: str
    platform: SocialPlatform
    content: str
    post_type: PostType
    media_urls: List[str] = Field(default_factory=list)
    status: PostStatus
    scheduled_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    engagement_metrics: Optional[Dict[str, Any]] = None
    hashtags: List[str] = Field(default_factory=list)
    mentions: List[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class CreatePostRequest(BaseModel):
    """Create social media post request"""
    account_ids: List[str]
    content: str
    post_type: PostType = PostType.TEXT
    media_urls: List[str] = Field(default_factory=list)
    scheduled_at: Optional[datetime] = None
    hashtags: List[str] = Field(default_factory=list)
    mentions: List[str] = Field(default_factory=list)


class UpdatePostRequest(BaseModel):
    """Update social media post request"""
    content: Optional[str] = None
    post_type: Optional[PostType] = None
    media_urls: Optional[List[str]] = None
    scheduled_at: Optional[datetime] = None
    hashtags: Optional[List[str]] = None
    mentions: Optional[List[str]] = None


class SocialAnalyticsResponse(BaseModel):
    """Social media analytics response"""
    total_posts: int
    posts_by_platform: Dict[str, int]
    posts_by_status: Dict[str, int]
    total_engagement: int
    average_engagement_rate: float
    top_performing_posts: List[SocialPostResponse]
    engagement_trend: List[Dict[str, Any]]
    follower_growth: List[Dict[str, Any]]
    best_posting_times: Dict[str, List[int]]
    
    model_config = ConfigDict(from_attributes=True)


class PostAnalyticsResponse(BaseModel):
    """Individual post analytics response"""
    post_id: str
    platform: SocialPlatform
    impressions: int
    reach: int
    likes: int
    comments: int
    shares: int
    saves: int
    clicks: int
    engagement_rate: float
    created_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class SocialQueueResponse(BaseModel):
    """Social media queue response"""
    scheduled_posts: List[SocialPostResponse]
    failed_posts: List[SocialPostResponse]
    next_post: Optional[SocialPostResponse] = None
    queue_length: int
    
    model_config = ConfigDict(from_attributes=True)

File: real-estate-dashboard/src/api/social.ts
/**
 * Social Media API Client
 */
import { apiClient } from '../lib/api/client';

export interface SocialAccount {
  id: string;
  platform: 'facebook' | 'instagram' | 'linkedin' | 'twitter' | 'youtube' | 'tiktok';
  account_name: string;
  account_id: string;
  access_token?: string;
  refresh_token?: string;
  is_active: boolean;
  status: 'active' | 'inactive' | 'error';
  last_sync?: string;
  follower_count?: number;
  following_count?: number;
  post_count?: number;
  engagement_rate?: number;
  created_at: string;
  updated_at: string;
}

export interface SocialPost {
  id: string;
  content: string;
  platform: string;
  post_type: 'text' | 'image' | 'video' | 'carousel' | 'story' | 'reel';
  status: 'draft' | 'scheduled' | 'published' | 'failed' | 'cancelled';
  scheduled_at?: string;
  published_at?: string;
  media_urls?: string[];
  hashtags?: string[];
  mentions?: string[];
  engagement?: {
    likes: number;
    comments: number;
    shares: number;
    clicks: number;
  };
  created_at: string;
  updated_at: string;
}

export interface SocialTemplate {
  id: string;
  name: string;
  template_type: string;
  content_template: string;
  hashtags: string[];
  platforms: string[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface SocialAnalytics {
  total_posts: number;
  total_engagements: number;
  total_reach: number;
  total_impressions: number;
  engagement_rate: number;
  platforms: Array<{
    platform: string;
    posts: number;
    engagements: number;
    reach: number;
    impressions: number;
    engagement_rate: number;
    followers: number;
    growth: number;
  }>;
  top_posts: Array<{
    id: string;
    content: string;
    platform: string;
    engagements: number;
    reach: number;
    impressions: number;
    engagement_rate: number;
    published_at: string;
  }>;
  audience: {
    demographics: {
      age: Record<string, number>;
      gender: Record<string, number>;
      location: Record<string, number>;
    };
    interests: Array<{
      name: string;
      percentage: number;
    }>;
    growth: {
      followers: number;
      following: number;
      posts: number;
    };
  };
  trends: {
    posts: Array<{ date: string; value: number }>;
    engagements: Array<{ date: string; value: number }>;
    reach: Array<{ date: string; value: number }>;
    impressions: Array<{ date: string; value: number }>;
  };
  start_date: string;
  end_date: string;
}

export interface CreatePostRequest {
  account_ids: string[];
  content: string;
  post_type: 'text' | 'image' | 'video' | 'carousel' | 'story' | 'reel';
  scheduled_at?: string;
  media_urls?: string[];
  hashtags?: string[];
  mentions?: string[];
}

export interface CreateTemplateRequest {
  name: string;
  template_type: string;
  content_template: string;
  hashtags?: string[];
  platforms?: string[];
}

export interface UpdateTemplateRequest {
  name?: string;
  template_type?: string;
  content_template?: string;
  hashtags?: string[];
  platforms?: string[];
  is_active?: boolean;
}

export interface SocialActivity {
  id: string;
  type: 'post_published' | 'post_scheduled' | 'post_created' | 'account_connected';
  title: string;
  description: string;
  time_ago: string;
  platform: string;
  post_id?: string;
  account_id?: string;
  created_at: string;
}

export interface SocialActivityResponse {
  activities: SocialActivity[];
}

export interface OAuthResponse {
  auth_url: string;
}

export interface PublishPostRequest {
  platforms: string[];
}

export interface SchedulePostRequest {
  platforms: string[];
  scheduled_at: string;
}

export const socialApi = {
  // Accounts
  getAccounts: () => apiClient.get<SocialAccount[]>('/api/v1/social/accounts'),
  
  connectAccount: (platform: string, token: string, accountId: string) =>
    apiClient.post<SocialAccount>('/api/v1/social/accounts', {
      platform,
      token,
      account_id: accountId
    }),
  
  disconnectAccount: (accountId: string) =>
    apiClient.delete(`/api/v1/social/accounts/${accountId}`),
  
  refreshAccountToken: (accountId: string) =>
    apiClient.post(`/api/v1/social/accounts/${accountId}/refresh`),

  // OAuth
  startOAuthFlow: (platform: string) =>
    apiClient.post<{ auth_url: string }>(`/api/v1/social/oauth/${platform}/authorize`),
  
  handleOAuthCallback: (platform: string, code: string, state: string) =>
    apiClient.get<SocialAccount>(`/api/v1/social/oauth/${platform}/callback`, {
      params: { code, state }
    }),
  
  // Account Management
  testAccountConnection: (accountId: string) =>
    apiClient.post<{ connected: boolean }>(`/api/v1/social/accounts/${accountId}/test`),
  
  syncAccountData: (accountId: string) =>
    apiClient.post<SocialAccount>(`/api/v1/social/accounts/${accountId}/sync`),
  
  // Post Publishing
  publishPostToPlatform: (postId: string, platform: string, accountId: string) =>
    apiClient.post(`/api/v1/social/posts/${postId}/publish/${platform}`, null, {
      params: { account_id: accountId }
    }),
  
  // Posts
  getPosts: (params?: {
    status?: string;
    platform?: string;
    limit?: number;
    offset?: number;
  }) => apiClient.get<SocialPost[]>('/api/v1/social/posts', { params }),
  
  createPost: (data: CreatePostRequest) =>
    apiClient.post<SocialPost>('/api/v1/social/posts', data),
  
  updatePost: (postId: string, data: Partial<CreatePostRequest>) =>
    apiClient.put<SocialPost>(`/api/v1/social/posts/${postId}`, data),
  
  deletePost: (postId: string) =>
    apiClient.delete(`/api/v1/social/posts/${postId}`),
  
  publishPost: (postId: string, data: PublishPostRequest) =>
    apiClient.post(`/api/v1/social/posts/${postId}/publish`, data),
  
  schedulePost: (postId: string, data: SchedulePostRequest) =>
    apiClient.post(`/api/v1/social/posts/${postId}/schedule`, data),

  // Templates
  getTemplates: (templateType?: string) =>
    apiClient.get<SocialTemplate[]>('/api/v1/social/templates', {
      params: templateType ? { template_type: templateType } : {}
    }),
  
  createTemplate: (data: CreateTemplateRequest) =>
    apiClient.post<SocialTemplate>('/api/v1/social/templates', data),
  
  getTemplate: (templateId: string) =>
    apiClient.get<SocialTemplate>(`/api/v1/social/templates/${templateId}`),
  
  updateTemplate: (templateId: string, data: UpdateTemplateRequest) =>
    apiClient.put<SocialTemplate>(`/api/v1/social/templates/${templateId}`, data),
  
  deleteTemplate: (templateId: string) =>
    apiClient.delete(`/api/v1/social/templates/${templateId}`),

  // Analytics
  getAnalytics: (params?: {
    start_date?: string;
    end_date?: string;
    platform?: string;
  }) => apiClient.get<SocialAnalytics>('/api/v1/social/analytics', { params }),

  // Queue
  getQueue: () => apiClient.get<SocialPost[]>('/api/v1/social/queue'),

  // OAuth
  startOAuth: (platform: string) =>
    apiClient.post<OAuthResponse>(`/api/v1/social/oauth/${platform}/authorize`),
  
  oauthCallback: (platform: string, code: string, state?: string) =>
    apiClient.get(`/api/v1/social/oauth/${platform}/callback`, {
      params: { code, state }
    }),
  
  refreshOAuthToken: (platform: string, accountId: string) =>
    apiClient.post(`/api/v1/social/oauth/${platform}/refresh`, { account_id: accountId }),

  // Media
  uploadMedia: (file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return apiClient.post('/api/v1/social/media', formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
  },

  // Activities
  getActivities: (limit: number = 10) =>
    apiClient.get<SocialActivity[]>(`/api/v1/social/activities?limit=${limit}`)
};

export default socialApi;
